daliuge-engine/dlg/drop.py

```diff
@@ -1591,10 +1598,15 @@ def cancel(self):
     def skip(self):
         '''Moves this application drop to its SKIPPED state'''
         super().skip()
+
+        prev_execStatus = self.execStatus
         self.execStatus = AppDROPStates.SKIPPED
         for o in self._outputs.values():
             o.skip()
-        self._fire('producerFinished', status=self.status, execStatus=self.execStatus)
+
+        logger.debug(f'Moving {self.__repr__()} to SKIPPED')
+        if prev_execStatus in [AppDROPStates.NOT_RUN]:
+            self._fire('producerFinished', status=self.status, execStatus=self.execStatus)
```

Comment:
AppDrops only need to signal "producerFinished" once in their lifetimes. Including the case for state "RUNNING" still causes multiple signals to occur as I assume a running node additionally changes to FINISHED state and signals again.  This condition might need to also apply for all transitions to finished states (`setCompleted()`, `setError()`, `cancel()`, `skipped()`) but to avoid complicated logic between `AppDROP` and `AbstractDROP` I've stuck with this for now.

Reply:
I guess ideally we'd have a concrete definition of this "state transition" logic. When the various events are generated etc.

I don't know how important it is for people working with DALiuGE to understand things at this level. Should we add it to some documentation somewhere, or just spread the knowledge internally. Or maybe just try to create a task to define this stuff with more detail.

This is just a comment and not anything actionable.